File: example/cpp/src/hardware/arm/free_drive.cpp
#include <cmath>
#include <iostream>
#include <thread>

#include "airbot_hardware/executors/executor.hpp"
#include "airbot_hardware/handlers/arm.hpp"

using namespace std::chrono_literals;
using namespace airbot::hardware;

bool all_zero(const std::array<double, 6>& pos) {
  for (const auto& p : pos)
    if (std::abs(p) > 1e-3) return false;
  return true;
}

int main() {
  auto executor = airbot::hardware::AsioExecutor::create(8);
  auto arm = airbot::hardware::Arm<6>::create<MotorType::OD, MotorType::OD, MotorType::OD, MotorType::DM, MotorType::DM,
                                              MotorType::DM, EEFType::NA, MotorType::NA>();
  if (!arm->init(executor->get_io_context(), "can0", 250)) {
    std::cerr << "Failed to initialize arm" << std::endl;
    return 1;
  }
  arm->enable();
  arm->set_param("arm.control_mode", static_cast<uint32_t>(MotorControlMode::MIT));
  // std::this_thread::sleep_for(1s);

  auto now = std::chrono::steady_clock::now();
  while (std::chrono::steady_clock::now() - now <= 10s) {
    arm->mit({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
             {0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0});
    std::this_thread::sleep_for(4ms);
  }

  arm->disable();
  arm->uninit();
  return 0;
}

File: example/python/hardware/arm/free_drive.py
import airbot_hardware_py as ah
import time


def main():
    executor = ah.create_asio_executor(8)
    io_context = executor.get_io_context()

    arm = ah.PlayWithEEF.create(
        ah.MotorType.EC,
        ah.MotorType.EC,
        ah.MotorType.EC,
        ah.MotorType.EC,
        ah.MotorType.EC,
        ah.MotorType.EC,
        ah.EEFType.E2,
        ah.MotorType.EC,
    )

    if not arm.init(io_context, "can0", 250):
        raise RuntimeError("The arm initialization failed.")
    print(arm.state().is_valid, arm.state().pos)

    arm.enable()
    arm.set_param("arm.control_mode", ah.MotorControlMode.MIT)
    now = time.time()
    while time.time() - now < 10:
        print(arm.state().is_valid, [f"{i:.4f}" for i in arm.state().pos])
        arm.mit(
            [0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0],
            [0, 0, 0, 0, 0, 0, 0],
        )
        time.sleep(0.004)

    arm.disable()
    arm.uninit()


if __name__ == "__main__":
    main()
